core/lib/server/pub.ts

```diff
@@ -589,3 +600,574 @@ export const deletePub = async (pubId: PubsId) =>
+	 * At time of writing (2024/11/04) I don't think that table has an `updatedAt`.
+	 */
+	updatedAt: Date;
+} & MaybePubChildren<Options> &
+	MaybePubStage<Options> &
+	MaybePubPubType<Options>;
+
+const DEFAULT_OPTIONS = {
+	depth: 2,
+	withChildren: true,
+	withRelatedPubs: true,
+	withPubType: false,
+	withStage: false,
+	cycle: "include",
+} as const satisfies GetPubsWithRelatedValuesAndChildrenOptions;
+
+export async function getPubsWithRelatedValuesAndChildren<
+	Options extends GetPubsWithRelatedValuesAndChildrenOptions,
+>(
+	props: {
+		pubId: PubsId;
+	},
+	options?: Options
+): Promise<ProcessedPub<Options>>;
+export async function getPubsWithRelatedValuesAndChildren<
+	Options extends GetPubsWithRelatedValuesAndChildrenOptions,
+>(
+	props: Exclude<PubIdOrPubTypeIdOrStageIdOrCommunityId, { pubId: PubsId }>,
+	options?: Options
+): Promise<ProcessedPub<Options>[]>;
+/**
+ * Retrieves a pub and all its related values, children, and related pubs up to a given depth.
+ */
+export async function getPubsWithRelatedValuesAndChildren<
+	Options extends GetPubsWithRelatedValuesAndChildrenOptions,
+>(
+	props: PubIdOrPubTypeIdOrStageIdOrCommunityId,
+	options?: Options
+): Promise<ProcessedPub<Options> | ProcessedPub<Options>[]> {
+	const opts = {
+		...DEFAULT_OPTIONS,
+		...options,
+	};
+
+	const {
+		depth,
+		withChildren,
+		withRelatedPubs,
+		cycle,
+		fieldSlugs,
+		orderBy,
+		orderDirection,
+		limit,
+		offset,
+		search,
+		withPubType,
+		withStage,
+	} = opts;
+
+	if (depth < 1) {
+		throw new Error("Depth must be a positive number");
+	}
+
+	const result = await autoCache(
+		db
+			// this pub_tree CTE roughly returns an array like so
+			// [
+			// 	{ pubId: 1, rootId: 1, parentId: null, depth: 1, value: 'Some value', valueId: 1, relatedPubId: null},
+			//  { pubId: 1, rootId: 1, parentId: 1, depth: 1, value: 'Some relationship value', valueId: 2, relatedPubId: 3},
+			//  { pubId: 2, rootId: 1, parentId: 1, depth: 2, value: 'Some child value', valueId: 3, relatedPubId: null},
+			//  { pubId: 3, rootId: 1, parentId: 2, depth: 2, value: 'Some related value', valueId: 4, relatedPubId: null},
+			// ]
+			// so it's an array of length (pub + children + relatedPubs) * values,
+			// with information about their depth, parent, and pub they are related to
+			//
+			// we could instead only look for the related and child pubs and ignore the other values
+			// but this would mean we would later need to look up the values for each pub as a subquery
+			// this way, the only subqueries we need below are looking up information already stored in pub_tree,
+			// which is more efficient
+			.withRecursive("pub_tree", (cte) =>
+				cte
+					// we need to do this weird cast, because kysely does not support typing the selecting from a later CTE
+					// which is possible only in a with recursive query
+					.selectFrom("root_pubs_limited as p" as unknown as "pubs as p")
```

Comment:
looks weird, but it's possible to reference "later" CTEs in recursive CTEs in postgres, see https://www.postgresql.org/docs/current/sql-select.html#SQL-WITH

Reply:
`kysely` obviously doesn't type that correctly though, hence the cast